### PR TITLE
[WIP] pulseaudio: build with meson

### PR DESCRIPTION
**Pull request by worldofpeace**
##### Blocking issue
Fails to build on darwin due to upstream bug with meson.build's
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/issues/734

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS (fails)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by FRidh** [comment]
Isn't .99 a pre-release?

**Comment by worldofpeace** [comment]
I think maybe? Can't completely remember, if it is perhaps it could just be merged to unstable for testing or continue to be a draft.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pulseaudio libpulseaudio pulseaudioFull

**Comment by worldofpeace** [comment]
Seems the darwin build is broken with
```
ld: unknown option: -z
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
```

so it shouldn't use that for darwin.
But it looks like the meson things are assuming ELF
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/blob/v12.99.3/meson.build#L345

Should I condition it to only add those when it's linux?

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pulseaudio libpulseaudio pulseaudioFull

**Comment by adisbladis** [comment]
> Isn't .99 a pre-release?

Yes it's a pre-release. 

> if it is perhaps it could just be merged to unstable for testing or continue to be a draft.

It should continue to be a draft until it can be updated to the 13.0 release.

**Comment by worldofpeace** [comment]
Yep https://lists.freedesktop.org/archives/pulseaudio-discuss/2019-September/031364.html
Release coming soon I guess
>  If things are quiet, we'll roll a final release in a few days.

Though it seems linking for Darwin is broken. I don't think I know enough about macos and darwin to make a bug report to them.

**Comment by jtojnar** [comment]
Looking at the error

```
Undefined symbols for architecture x86_64:
  "_AbsoluteToNanoseconds", referenced from:
      _pa_rtclock_age in pulsecore_core-rtclock.c.o
      _pa_rtclock_get in pulsecore_core-rtclock.c.o
      _pa_rtclock_from_wallclock in pulsecore_core-rtclock.c.o
      _pa_timeval_rtstore in pulsecore_core-rtclock.c.o
  "_libintl_bind_textdomain_codeset", referenced from:
      _pa_init_i18n in pulsecore_i18n.c.o
  "_libintl_bindtextdomain", referenced from:
      _pa_init_i18n in pulsecore_i18n.c.o
  "_libintl_dgettext", referenced from:
      _pa_format_info_snprint in pulse_format.c.o
      _pa_channel_position_to_pretty_string in pulse_channelmap.c.o
      _pa_channel_map_snprint in pulse_channelmap.c.o
      _pa_channel_map_to_pretty_name in pulse_channelmap.c.o
      _pa_sample_spec_snprint in pulse_sample.c.o
      _pa_bytes_snprint in pulse_sample.c.o
      _pa_cvolume_snprint in pulse_volume.c.o
      ...
  "_libintl_newlocale", referenced from:
      _pa_atod in pulsecore_core-util.c.o
  "_libintl_setlocale", referenced from:
      _pa_init_proplist in pulsecore_proplist-util.c.o
  "_pa_poll", referenced from:
      _pa_autospawn_lock_acquire in pulsecore_lock-autospawn.c.o
```

and some web search points me to underlinking.

https://github.com/d5/node.native/issues/8#issuecomment-6379868

The libintl linking might be fixed by just adding `libintl` to `buildInputs` (it should be no-op on glibc), not sure if we also need to link it explicitly.

cc @NixOS/darwin-maintainers

**Comment by worldofpeace** [comment]
Perhaps we need a meson equivalent of doing
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/blob/master/configure.ac#L1404
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/blob/master/configure.ac#L536

**Comment by jtojnar** [comment]
Yes for the frameworks. For the `enable_mac_universal`, that probably refers to https://en.wikipedia.org/wiki/Universal_binary, I am not sure if that is still needed these days.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pulseaudio libpulseaudio pulseaudioFull

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pulseaudio

**Comment by jtojnar** [comment]
Possibly relevant https://gitlab.freedesktop.org/pulseaudio/pulseaudio/issues/707

**Comment by worldofpeace** [comment]
This is what fedora did to fix it
* https://src.fedoraproject.org/rpms/qt5-qtwebengine/c/4be2cf030b67b001b3fc7b733723ea12143f5c2a?branch=master
* https://src.fedoraproject.org/rpms/qt5-qtwebengine/c/d26d4f2572e5db0b5462d8cb5feba2221d420f10?branch=master

File: pkgs/servers/pulseaudio/default.nix [hunk 1]
```diff
@@ -114,38 +163,32 @@ stdenv.mkDerivation rec {
...
-      "pulseconfdir=$(out)/etc/pulse"
-    ];
+  NIX_LDFLAGS = lib.optionalString stdenv.isDarwin "-framework CoreServices -framework Cocoa -framework AudioUnit";
 
   postInstall = lib.optionalString libOnly ''
     rm -rf $out/{bin,share,etc,lib/{pulse-*,systemd}}
-    sed 's|-lltdl|-L${libtool.lib}/lib -lltdl|' -i $out/lib/pulseaudio/libpulsecore-${version}.la
-  ''
-    + ''moveToOutput lib/cmake "$dev" '';
+    moveToOutput lib/cmake "$dev"
```
Comment (jtojnar) [inline pkgs/servers/pulseaudio/default.nix]: Should not this also be done when `!libOnly`?

Comment (worldofpeace) [reply pkgs/servers/pulseaudio/default.nix]: did that without !libOnly

**Comment by worldofpeace** [comment]
v13.0 released
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/tags/v13.0

**Comment by worldofpeace** [comment]
Opened an issue about darwin and meson, although it's lackluster in information
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/issues/734

**Comment by jtojnar** [comment]
Notes on the website: https://www.freedesktop.org/wiki/Software/PulseAudio/Notes/13.0/

**Comment by worldofpeace** [comment]
I've confirmed that 
* https://gitlab.freedesktop.org/pulseaudio/pulseaudio/merge_requests/171

helped with the darwin build slightly.

**Review by ttuegel — APPROVED** [review]

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pulseaudio libpulseaudio pulseaudioFull

**Comment by worldofpeace** [comment]
I've merged https://github.com/NixOS/nixpkgs/pull/68735 for now, I don't want to break darwin.

Leaving this open until someone fixes the issue upstream.

**Comment by davidak** [comment]
Since worldofpeace is not here anymore:

Does someone want to continue this effort?

Or can we close this issue?

**Comment by Ericson2314** [comment]
Do they have plans to drop autoconf eventually?

**Comment by primeos** [comment]
@Ericson2314 yes, the recent PulseAudio 15 release dropped support for Autotools: https://www.freedesktop.org/wiki/Software/PulseAudio/Notes/15.0/#autotoolsbuildsystemhavebeendropped

So we need to switch to Meson to update PulseAudio 15. Is someone up for that / already working on it?
(If not I might give it a shot soon but I'm not sure if I have enough time for it.)

**Pull request closed without merging**